compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/TypeDefBuilderHelper.java

```diff
@@ -220,15 +221,33 @@ public static BLangClassDefinition createClassDef(Location pos, BObjectTypeSymbo
         return classDefNode;
     }
 
-    public static BLangErrorType createBLangErrorType(Location pos, BType detailType) {
+    public static BLangErrorType createBLangErrorType(Location pos, BType detailType, SymbolEnv env, Names names) {
         BLangErrorType errorType = (BLangErrorType) TreeBuilder.createErrorTypeNode();
         BLangUserDefinedType userDefinedTypeNode = (BLangUserDefinedType) TreeBuilder.createUserDefinedTypeNode();
         userDefinedTypeNode.pos = pos;
         userDefinedTypeNode.pkgAlias = (BLangIdentifier) TreeBuilder.createIdentifierNode();
-        userDefinedTypeNode.typeName = createIdentifier(pos, detailType.tsymbol.name.value);
+
+        // When the error detail type is a inline type decl, such as in `error<map<value:Cloneable>>` we need to
+        // add a explicit type decl for the detail type.
+        boolean userDefinedDetailTypeAvailable = detailType.tsymbol != null;
+        String typeName = userDefinedDetailTypeAvailable
+                ? detailType.tsymbol.name.value
+                : "$synthetic$error$detail$" + errorDetailAnnonCount++;
+
+        userDefinedTypeNode.typeName = createIdentifier(pos, typeName);
         userDefinedTypeNode.type = detailType;
         errorType.detailType = userDefinedTypeNode;
 
+        // Add explicit type definition for this error detail
+        if (!userDefinedDetailTypeAvailable) {
+            BTypeSymbol typeSymbol = new BTypeSymbol(SymTag.TYPE, Flags.ANONYMOUS,
+                    names.fromString(typeName), env.enclPkg.packageID, userDefinedTypeNode.type,
+                    env.enclPkg.symbol, pos, VIRTUAL);
+            addTypeDefinition(userDefinedTypeNode.type, typeSymbol, userDefinedTypeNode, env);
+            env.enclPkg.symbol.scope.define(typeSymbol.name, typeSymbol);
+            userDefinedTypeNode.type.tsymbol = typeSymbol;
+        }
```

Comment:
Don't we already add type definitions for anon types (at desugar for local anon record types IIRC) irrespective of where they are?

Reply:
When the detail type is a map, there is no user-defined type, that's what being defined here